File: tests/algorithms/x/Pascal/ciphers/base85.pas
{$mode objfpc}
program Main;
uses SysUtils;
var _nowSeed: int64 = 0;
var _nowSeeded: boolean = false;
procedure init_now();
var s: string; v: int64;
begin
  s := GetEnvironmentVariable('MOCHI_NOW_SEED');
  if s <> '' then begin
    Val(s, v);
    _nowSeed := v;
    _nowSeeded := true;
  end;
end;
function _now(): integer;
begin
  if _nowSeeded then begin
    _nowSeed := (_nowSeed * 1664525 + 1013904223) mod 2147483647;
    _now := _nowSeed;
  end else begin
    _now := Integer(GetTickCount64()*1000);
  end;
end;
function _bench_now(): int64;
begin
  _bench_now := GetTickCount64()*1000;
end;
function _mem(): int64;
var h: TFPCHeapStatus;
begin
  h := GetFPCHeapStatus;
  _mem := h.CurrHeapUsed;
end;
procedure panic(msg: string);
begin
  writeln(msg);
  halt(1);
end;
var
  bench_start_0: integer;
  bench_dur_0: integer;
  bench_mem_0: int64;
  bench_memdiff_0: int64;
  ascii85_chars: string;
  s: string;
  digits: string;
  n: integer;
  ch: string;
  data: string;
  d: integer;
  bits: string;
function indexOf(s: string; ch: string): integer; forward;
function ord(ch: string): integer; forward;
function chr(n: integer): string; forward;
function to_binary(n: integer; bits: integer): string; forward;
function bin_to_int(bits: string): integer; forward;
function reverse(s: string): string; forward;
function base10_to_85(d: integer): string; forward;
function base85_to_10(digits: string): integer; forward;
function ascii85_encode(data: string): string; forward;
function ascii85_decode(data: string): string; forward;
function indexOf(s: string; ch: string): integer;
var
  indexOf_i: integer;
begin
  indexOf_i := 0;
  while indexOf_i < Length(s) do begin
  if s[indexOf_i+1] = ch then begin
  exit(indexOf_i);
end;
  indexOf_i := indexOf_i + 1;
end;
  exit(-1);
end;
function ord(ch: string): integer;
var
  ord_idx: integer;
begin
  ord_idx := indexOf(ascii85_chars, ch);
  if ord_idx >= 0 then begin
  exit(33 + ord_idx);
end;
  exit(0);
end;
function chr(n: integer): string;
begin
  if (n >= 33) and (n <= 117) then begin
  exit(copy(ascii85_chars, n - 33+1, (n - 32 - (n - 33))));
end;
  exit('?');
end;
function to_binary(n: integer; bits: integer): string;
var
  to_binary_b: string;
  to_binary_val: integer;
begin
  to_binary_b := '';
  to_binary_val := n;
  while to_binary_val > 0 do begin
  to_binary_b := IntToStr(to_binary_val mod 2) + to_binary_b;
  to_binary_val := to_binary_val div 2;
end;
  while Length(to_binary_b) < bits do begin
  to_binary_b := '0' + to_binary_b;
end;
  if Length(to_binary_b) = 0 then begin
  to_binary_b := '0';
end;
  exit(to_binary_b);
end;
function bin_to_int(bits: string): integer;
var
  bin_to_int_n: integer;
  bin_to_int_i: integer;
begin
  bin_to_int_n := 0;
  bin_to_int_i := 0;
  while bin_to_int_i < Length(bits) do begin
  if bits[bin_to_int_i+1] = '1' then begin
  bin_to_int_n := (bin_to_int_n * 2) + 1;
end else begin
  bin_to_int_n := bin_to_int_n * 2;
end;
  bin_to_int_i := bin_to_int_i + 1;
end;
  exit(bin_to_int_n);
end;
function reverse(s: string): string;
var
  reverse_res: string;
  reverse_i: integer;
begin
  reverse_res := '';
  reverse_i := Length(s) - 1;
  while reverse_i >= 0 do begin
  reverse_res := reverse_res + s[reverse_i+1];
  reverse_i := reverse_i - 1;
end;
  exit(reverse_res);
end;
function base10_to_85(d: integer): string;
begin
  if d > 0 then begin
  exit(chr((d mod 85) + 33) + base10_to_85(d div 85));
end;
  exit('');
end;
function base85_to_10(digits: string): integer;
var
  base85_to_10_value: integer;
  base85_to_10_i: integer;
begin
  base85_to_10_value := 0;
  base85_to_10_i := 0;
  while base85_to_10_i < Length(digits) do begin
  base85_to_10_value := (base85_to_10_value * 85) + (ord(digits[base85_to_10_i+1]) - 33);
  base85_to_10_i := base85_to_10_i + 1;
end;
  exit(base85_to_10_value);
end;
function ascii85_encode(data: string): string;
var
  ascii85_encode_binary_data: string;
  ascii85_encode_ch: integer;
  ascii85_encode_null_values: integer;
  ascii85_encode_total_bits: integer;
  ascii85_encode_result_: string;
  ascii85_encode_i: integer;
  ascii85_encode_chunk_bits: string;
  ascii85_encode_chunk_val: integer;
  ascii85_encode_encoded: string;
begin
  ascii85_encode_binary_data := '';
  for ascii85_encode_ch in data do begin
  ascii85_encode_binary_data := ascii85_encode_binary_data + to_binary(ord(ascii85_encode_ch), 8);
end;
  ascii85_encode_null_values := ((32 * ((Length(ascii85_encode_binary_data) div 32) + 1)) - Length(ascii85_encode_binary_data)) div 8;
  ascii85_encode_total_bits := 32 * ((Length(ascii85_encode_binary_data) div 32) + 1);
  while Length(ascii85_encode_binary_data) < ascii85_encode_total_bits do begin
  ascii85_encode_binary_data := ascii85_encode_binary_data + '0';
end;
  ascii85_encode_result_ := '';
  ascii85_encode_i := 0;
  while ascii85_encode_i < Length(ascii85_encode_binary_data) do begin
  ascii85_encode_chunk_bits := copy(ascii85_encode_binary_data, ascii85_encode_i+1, (ascii85_encode_i + 32 - (ascii85_encode_i)));
  ascii85_encode_chunk_val := bin_to_int(ascii85_encode_chunk_bits);
  ascii85_encode_encoded := reverse(base10_to_85(ascii85_encode_chunk_val));
  ascii85_encode_result_ := ascii85_encode_result_ + ascii85_encode_encoded;
  ascii85_encode_i := ascii85_encode_i + 32;
end;
  if (ascii85_encode_null_values mod 4) <> 0 then begin
  ascii85_encode_result_ := copy(ascii85_encode_result_, 0+1, (Length(ascii85_encode_result_) - ascii85_encode_null_values - (0)));
end;
  exit(ascii85_encode_result_);
end;
function ascii85_decode(data: string): string;
var
  ascii85_decode_null_values: integer;
  ascii85_decode_binary_data: string;
  ascii85_decode_i: integer;
  ascii85_decode_result_: string;
  ascii85_decode_chunk: string;
  ascii85_decode_value: integer;
  ascii85_decode_bits: string;
  ascii85_decode_j: integer;
  ascii85_decode_byte_bits: string;
  ascii85_decode_c: string;
  ascii85_decode_trim: integer;
begin
  ascii85_decode_null_values := (5 * ((Length(data) div 5) + 1)) - Length(data);
  ascii85_decode_binary_data := data;
  ascii85_decode_i := 0;
  while ascii85_decode_i < ascii85_decode_null_values do begin
  ascii85_decode_binary_data := ascii85_decode_binary_data + 'u';
  ascii85_decode_i := ascii85_decode_i + 1;
end;
  ascii85_decode_result_ := '';
  ascii85_decode_i := 0;
  while ascii85_decode_i < Length(ascii85_decode_binary_data) do begin
  ascii85_decode_chunk := copy(ascii85_decode_binary_data, ascii85_decode_i+1, (ascii85_decode_i + 5 - (ascii85_decode_i)));
  ascii85_decode_value := base85_to_10(ascii85_decode_chunk);
  ascii85_decode_bits := to_binary(ascii85_decode_value, 32);
  ascii85_decode_j := 0;
  while ascii85_decode_j < 32 do begin
  ascii85_decode_byte_bits := copy(ascii85_decode_bits, ascii85_decode_j+1, (ascii85_decode_j + 8 - (ascii85_decode_j)));
  ascii85_decode_c := chr(bin_to_int(ascii85_decode_byte_bits));
  ascii85_decode_result_ := ascii85_decode_result_ + ascii85_decode_c;
  ascii85_decode_j := ascii85_decode_j + 8;
end;
  ascii85_decode_i := ascii85_decode_i + 5;
end;
  ascii85_decode_trim := ascii85_decode_null_values;
  if (ascii85_decode_null_values mod 5) = 0 then begin
  ascii85_decode_trim := ascii85_decode_null_values - 1;
end;
  exit(copy(ascii85_decode_result_, 0+1, (Length(ascii85_decode_result_) - ascii85_decode_trim - (0))));
end;
begin
  init_now();
  bench_mem_0 := _mem();
  bench_start_0 := _bench_now();
  ascii85_chars := '!"#$%&''()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\]^_`abcdefghijklmnopqrstu';
  writeln(ascii85_encode(''));
  writeln(ascii85_encode('12345'));
  writeln(ascii85_encode('base 85'));
  writeln(ascii85_decode(''));
  writeln(ascii85_decode('0etOA2#'));
  writeln(ascii85_decode('@UX=h+?24'));
  bench_memdiff_0 := _mem() - bench_mem_0;
  bench_dur_0 := (_bench_now() - bench_start_0) div 1000;
  writeln('{');
  writeln(('  "duration_us": ' + IntToStr(bench_dur_0)) + ',');
  writeln(('  "memory_bytes": ' + IntToStr(bench_memdiff_0)) + ',');
  writeln(('  "name": "' + 'main') + '"');
  writeln('}');
end.

File: tests/algorithms/x/Clojure/ciphers/base85.clj
(ns main (:refer-clojure :exclude [indexOf ord chr to_binary bin_to_int reverse base10_to_85 base85_to_10 ascii85_encode ascii85_decode]))

(require 'clojure.set)

(defn in [x coll]
  (cond (string? coll) (clojure.string/includes? coll x) (map? coll) (contains? coll x) (sequential? coll) (some (fn [e] (= e x)) coll) :else false))

(defn padStart [s w p]
  (loop [out (str s)] (if (< (count out) w) (recur (str p out)) out)))

(defn indexOf [s sub]
  (let [idx (clojure.string/index-of s sub)] (if (nil? idx) -1 idx)))

(defn split [s sep]
  (clojure.string/split s (re-pattern sep)))

(def nowSeed (atom (let [s (System/getenv "MOCHI_NOW_SEED")] (if (and s (not (= s ""))) (Integer/parseInt s) 0))))

(declare indexOf ord chr to_binary bin_to_int reverse base10_to_85 base85_to_10 ascii85_encode ascii85_decode)

(def ^:dynamic ascii85_decode_binary_data nil)

(def ^:dynamic ascii85_decode_bits nil)

(def ^:dynamic ascii85_decode_byte_bits nil)

(def ^:dynamic ascii85_decode_c nil)

(def ^:dynamic ascii85_decode_chunk nil)

(def ^:dynamic ascii85_decode_i nil)

(def ^:dynamic ascii85_decode_j nil)

(def ^:dynamic ascii85_decode_null_values nil)

(def ^:dynamic ascii85_decode_result nil)

(def ^:dynamic ascii85_decode_trim nil)

(def ^:dynamic ascii85_decode_value nil)

(def ^:dynamic ascii85_encode_binary_data nil)

(def ^:dynamic ascii85_encode_chunk_bits nil)

(def ^:dynamic ascii85_encode_chunk_val nil)

(def ^:dynamic ascii85_encode_encoded nil)

(def ^:dynamic ascii85_encode_i nil)

(def ^:dynamic ascii85_encode_null_values nil)

(def ^:dynamic ascii85_encode_result nil)

(def ^:dynamic ascii85_encode_total_bits nil)

(def ^:dynamic base85_to_10_i nil)

(def ^:dynamic base85_to_10_value nil)

(def ^:dynamic bin_to_int_i nil)

(def ^:dynamic bin_to_int_n nil)

(def ^:dynamic indexOf_i nil)

(def ^:dynamic ord_idx nil)

(def ^:dynamic reverse_i nil)

(def ^:dynamic reverse_res nil)

(def ^:dynamic to_binary_b nil)

(def ^:dynamic to_binary_val nil)

(def ^:dynamic main_ascii85_chars "!\"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstu")

(defn indexOf [indexOf_s indexOf_ch]
  (binding [indexOf_i nil] (try (do (set! indexOf_i 0) (while (< indexOf_i (count indexOf_s)) (do (when (= (nth indexOf_s indexOf_i) indexOf_ch) (throw (ex-info "return" {:v indexOf_i}))) (set! indexOf_i (+ indexOf_i 1)))) (throw (ex-info "return" {:v (- 1)}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e))))))

(defn ord [ord_ch]
  (binding [ord_idx nil] (try (do (set! ord_idx (indexOf main_ascii85_chars ord_ch)) (if (>= ord_idx 0) (+ 33 ord_idx) 0)) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e))))))

(defn chr [chr_n]
  (try (if (and (>= chr_n 33) (<= chr_n 117)) (subs main_ascii85_chars (- chr_n 33) (min (- chr_n 32) (count main_ascii85_chars))) "?") (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e)))))

(defn to_binary [to_binary_n to_binary_bits]
  (binding [to_binary_b nil to_binary_val nil] (try (do (set! to_binary_b "") (set! to_binary_val to_binary_n) (while (> to_binary_val 0) (do (set! to_binary_b (str (str (mod to_binary_val 2)) to_binary_b)) (set! to_binary_val (quot to_binary_val 2)))) (while (< (count to_binary_b) to_binary_bits) (set! to_binary_b (str "0" to_binary_b))) (when (= (count to_binary_b) 0) (set! to_binary_b "0")) (throw (ex-info "return" {:v to_binary_b}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e))))))

(defn bin_to_int [bin_to_int_bits]
  (binding [bin_to_int_i nil bin_to_int_n nil] (try (do (set! bin_to_int_n 0) (set! bin_to_int_i 0) (while (< bin_to_int_i (count bin_to_int_bits)) (do (if (= (nth bin_to_int_bits bin_to_int_i) "1") (set! bin_to_int_n (+ (* bin_to_int_n 2) 1)) (set! bin_to_int_n (* bin_to_int_n 2))) (set! bin_to_int_i (+ bin_to_int_i 1)))) (throw (ex-info "return" {:v bin_to_int_n}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e))))))

(defn reverse [reverse_s]
  (binding [reverse_i nil reverse_res nil] (try (do (set! reverse_res "") (set! reverse_i (- (count reverse_s) 1)) (while (>= reverse_i 0) (do (set! reverse_res (str reverse_res (nth reverse_s reverse_i))) (set! reverse_i (- reverse_i 1)))) (throw (ex-info "return" {:v reverse_res}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e))))))

(defn base10_to_85 [base10_to_85_d]
  (try (if (> base10_to_85_d 0) (str (chr (+ (mod base10_to_85_d 85) 33)) (base10_to_85 (quot base10_to_85_d 85))) "") (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e)))))

(defn base85_to_10 [base85_to_10_digits]
  (binding [base85_to_10_i nil base85_to_10_value nil] (try (do (set! base85_to_10_value 0) (set! base85_to_10_i 0) (while (< base85_to_10_i (count base85_to_10_digits)) (do (set! base85_to_10_value (+ (* base85_to_10_value 85) (- (ord (nth base85_to_10_digits base85_to_10_i)) 33))) (set! base85_to_10_i (+ base85_to_10_i 1)))) (throw (ex-info "return" {:v base85_to_10_value}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e))))))

(defn ascii85_encode [ascii85_encode_data]
  (binding [ascii85_encode_binary_data nil ascii85_encode_chunk_bits nil ascii85_encode_chunk_val nil ascii85_encode_encoded nil ascii85_encode_i nil ascii85_encode_null_values nil ascii85_encode_result nil ascii85_encode_total_bits nil] (try (do (set! ascii85_encode_binary_data "") (doseq [ch ascii85_encode_data] (set! ascii85_encode_binary_data (str ascii85_encode_binary_data (to_binary (ord ch) 8)))) (set! ascii85_encode_null_values (quot (- (* 32 (+ (quot (count ascii85_encode_binary_data) 32) 1)) (count ascii85_encode_binary_data)) 8)) (set! ascii85_encode_total_bits (* 32 (+ (quot (count ascii85_encode_binary_data) 32) 1))) (while (< (count ascii85_encode_binary_data) ascii85_encode_total_bits) (set! ascii85_encode_binary_data (str ascii85_encode_binary_data "0"))) (set! ascii85_encode_result "") (set! ascii85_encode_i 0) (while (< ascii85_encode_i (count ascii85_encode_binary_data)) (do (set! ascii85_encode_chunk_bits (subs ascii85_encode_binary_data ascii85_encode_i (min (+ ascii85_encode_i 32) (count ascii85_encode_binary_data)))) (set! ascii85_encode_chunk_val (bin_to_int ascii85_encode_chunk_bits)) (set! ascii85_encode_encoded (reverse (base10_to_85 ascii85_encode_chunk_val))) (set! ascii85_encode_result (str ascii85_encode_result ascii85_encode_encoded)) (set! ascii85_encode_i (+ ascii85_encode_i 32)))) (when (not= (mod ascii85_encode_null_values 4) 0) (set! ascii85_encode_result (subs ascii85_encode_result 0 (min (- (count ascii85_encode_result) ascii85_encode_null_values) (count ascii85_encode_result))))) (throw (ex-info "return" {:v ascii85_encode_result}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e))))))

(defn ascii85_decode [ascii85_decode_data]
  (binding [ascii85_decode_binary_data nil ascii85_decode_bits nil ascii85_decode_byte_bits nil ascii85_decode_c nil ascii85_decode_chunk nil ascii85_decode_i nil ascii85_decode_j nil ascii85_decode_null_values nil ascii85_decode_result nil ascii85_decode_trim nil ascii85_decode_value nil] (try (do (set! ascii85_decode_null_values (- (* 5 (+ (quot (count ascii85_decode_data) 5) 1)) (count ascii85_decode_data))) (set! ascii85_decode_binary_data ascii85_decode_data) (set! ascii85_decode_i 0) (while (< ascii85_decode_i ascii85_decode_null_values) (do (set! ascii85_decode_binary_data (str ascii85_decode_binary_data "u")) (set! ascii85_decode_i (+ ascii85_decode_i 1)))) (set! ascii85_decode_result "") (set! ascii85_decode_i 0) (while (< ascii85_decode_i (count ascii85_decode_binary_data)) (do (set! ascii85_decode_chunk (subs ascii85_decode_binary_data ascii85_decode_i (min (+ ascii85_decode_i 5) (count ascii85_decode_binary_data)))) (set! ascii85_decode_value (base85_to_10 ascii85_decode_chunk)) (set! ascii85_decode_bits (to_binary ascii85_decode_value 32)) (set! ascii85_decode_j 0) (while (< ascii85_decode_j 32) (do (set! ascii85_decode_byte_bits (subs ascii85_decode_bits ascii85_decode_j (min (+ ascii85_decode_j 8) (count ascii85_decode_bits)))) (set! ascii85_decode_c (chr (bin_to_int ascii85_decode_byte_bits))) (set! ascii85_decode_result (str ascii85_decode_result ascii85_decode_c)) (set! ascii85_decode_j (+ ascii85_decode_j 8)))) (set! ascii85_decode_i (+ ascii85_decode_i 5)))) (set! ascii85_decode_trim ascii85_decode_null_values) (when (= (mod ascii85_decode_null_values 5) 0) (set! ascii85_decode_trim (- ascii85_decode_null_values 1))) (throw (ex-info "return" {:v (subs ascii85_decode_result 0 (min (- (count ascii85_decode_result) ascii85_decode_trim) (count ascii85_decode_result)))}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e))))))

(defn -main []
  (let [rt (Runtime/getRuntime)
    start-mem (- (.totalMemory rt) (.freeMemory rt))
    start (System/nanoTime)]
      (println (ascii85_encode ""))
      (println (ascii85_encode "12345"))
      (println (ascii85_encode "base 85"))
      (println (ascii85_decode ""))
      (println (ascii85_decode "0etOA2#"))
      (println (ascii85_decode "@UX=h+?24"))
      (System/gc)
      (let [end (System/nanoTime)
        end-mem (- (.totalMemory rt) (.freeMemory rt))
        duration-us (quot (- end start) 1000)
        memory-bytes (Math/abs ^long (- end-mem start-mem))]
        (println (str "{\n  \"duration_us\": " duration-us ",\n  \"memory_bytes\": " memory-bytes ",\n  \"name\": \"main\"\n}"))
      )
    ))

(-main)
